core/output/smartcard.js

```diff
@@ -0,0 +1,318 @@
+import * as Log from "../../core/util/logging.js";
+import * as WebUtil from "../../app/webutil.js";
+
+// utilities
+const toHex = (data) => {
+  return Array.from(data)
+    .map((b) => b.toString(16).padStart(2, "0").toUpperCase())
+    .join("")
+    .replace(/ /g, "");
+};
+
+const fromHex = (data) => {
+  return new Uint8Array(data.match(/.{1,2}/g).map((b) => parseInt(b, 16)));
```

Comment:
this will likely throw an exception if an empty string is passed, likely need input validation unless it is not possible, but even if it is not possible should likely guard against it and handle appropriately.

Reply:
@mmcclaskey I updated the function to handle empty values and made the exceptions more explicit too. I prefer not to fail silently here because it could lead to a very subtle bugs (happened to me when I was working on the guacd version).